src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,169 @@
+    {
+        protected override IAnalyzer CreateAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+            => new CSharpAnalyzer(syntaxFacts, semanticModel);
+
+        private sealed class CSharpAnalyzer : Analyzer<QueryExpressionSyntax, ExpressionSyntax>
+        {
+            public CSharpAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+                : base(syntaxFacts, semanticModel)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_linq_query_to_linq_method;
+
+            protected override ExpressionSyntax Convert(QueryExpressionSyntax source)
+            {
+                var fromClause = source.FromClause;
+                var identifier = fromClause.Identifier;
+                var expression = fromClause.Expression.WithoutTrailingTrivia();
+                if (fromClause.Type != null)
+                {
+                    var lambda = SyntaxFactory.BinaryExpression(SyntaxKind.IsExpression, SyntaxFactory.IdentifierName(identifier), SyntaxFactory.Token(SyntaxKind.IsKeyword), fromClause.Type);
+                    expression = CreateInvocationExpression(expression, identifier, lambda, "Select");
+                }
+
+                var context = new ConversionContext(identifier, expression);
+                return ProcessQueryBody(context, source.Body);
+            }
+
+            protected override bool Validate(QueryExpressionSyntax originalMethod, ExpressionSyntax convertedMethod, CancellationToken cancellationToken)
+            {
+                var speculationAnalyzer = new SpeculationAnalyzer(originalMethod, convertedMethod, _semanticModel, cancellationToken);
+                if (speculationAnalyzer.ReplacementChangesSemantics())
+                {
+                    return false;
+                }
+
+                // TODO add more checks
+                return true;
+            }
+
+            private static ExpressionSyntax ProcessQueryBody(ConversionContext context, QueryBodySyntax queryBody)
+            {
+                var expression = context.Expression;
+                foreach (var queryClause in queryBody.Clauses)
+                {
+                    switch (queryClause.Kind())
+                    {
+                        case SyntaxKind.WhereClause: expression = CreateInvocationExpression(expression, context.Identifier, ((WhereClauseSyntax)queryClause).Condition, "Where"); break;
+                        case SyntaxKind.OrderByClause: expression = ProcessOrderByClause(expression, context.Identifier, (OrderByClauseSyntax)queryClause); break;
+                        case SyntaxKind.JoinClause: // Not supported because linq queries seem to provide essentially simpler syntax for the same than query methods.
+                        case SyntaxKind.FromClause: // More than one fromClause is not supported. The linq method seems to be more complicated for this.
```

Comment:
since this is a refactoring, i think it would make sense to still offer.  The user can decide if the conversion is ok.

Reply:
OK. [https://github.com/dotnet/roslyn/issues/25112](https://github.com/dotnet/roslyn/issues/25112)

---
In reply to: [167387556](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387556) [](ancestors = 167387556)